packages/zk-certificates/contracts/AirdropGateway.sol

```diff
@@ -0,0 +1,137 @@
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "./interfaces/IZkKYCVerifier.sol";
+
+/**
+ * @title AirdropGateway: manages airdrops for the Galactica
+ */
+contract AirdropGateway is AccessControl {
+    // roles for access control
+    bytes32 public constant CLIENT_ROLE = keccak256("CLIENT_ROLE");
+    uint public distributionIndexCounter;
+    // due to some Solidity peculiarity, the associated getter doesn't return the dynamic array of requiredSBTs
+    mapping(uint => AirdropDistribution) public distributions;
+    // distributionIndex -> userAddress -> registration status
+    mapping(uint => mapping(address => bool)) public registeredUsers;
+    // distribution Index -> humanId => registration status
+    mapping(uint => mapping(bytes32 => bool)) public registeredHumanID;
+    // distributionIndex -> userAddress -> claimed status
+    mapping(uint => mapping(address => bool)) public claimedUsers;
+    IZkKYCVerifier public verifierWrapper;
+
+    event DistributionCreated(uint indexed distributionId, address client);
+    event UserRegistered(uint indexed distributionId, address user);
+    event UserClaimed(uint indexed distributionId, address user, uint amount);
+
+    // struct to store airdrop distribution information
+    struct AirdropDistribution {
+        address[] requiredSBTs;
+        uint registrationStartTime;
+        uint registrationEndTime;
+        uint claimStartTime;
+        uint claimEndTime;
+        address clientAddress;
+        address tokenAddress;
+        uint256 distributionAmount;
+        uint256 registeredUserCount;
+        uint256 tokenAmountPerUser;
+        uint256 amountClaimed;
+    }
+
+    constructor(address owner, address verifierWrapperAddress) {
+        verifierWrapper = IZkKYCVerifier(verifierWrapperAddress);
+        // set admin, the role that can assign and revoke other roles
+        _setupRole(DEFAULT_ADMIN_ROLE, owner);
+    }
+
+    function whitelistClient(address newClient) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _grantRole(CLIENT_ROLE, newClient);
+    }
+
+    function dewhitelistClient(address client) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _revokeRole(CLIENT_ROLE, client);
+    }
+
+    function setDistribution(address[] memory requiredSBTs, address tokenAddress, uint registrationStartTime, uint registrationEndTime, uint claimStartTime, uint claimEndTime) external onlyRole(CLIENT_ROLE) {
+        require(registrationStartTime < registrationEndTime, "invalid registration time");
+        require(registrationEndTime < claimStartTime, "claim can only start after registration ends");
+        require(claimStartTime < claimEndTime, "invalid claim time");
+        distributions[distributionIndexCounter] = AirdropDistribution(requiredSBTs, registrationStartTime, registrationEndTime, claimStartTime, claimEndTime, msg.sender, tokenAddress, 0, 0, 0, 0);
```

Comment:
Not sure if this is a stupid question:
Should the client also be required to be KYCed for compliance? Is this covered in the whitelisting process already?

Reply:
Since we, the admin, we be the ones calling this function, this KYC, if needed, can be done offchain.